[BPK-2731]: New configurable nudger component (#1553)
Skyscanner/backpack: packages/bpk-component-nudger/README.md

## Usage

> ### Default `BpkNudger`

Reviewer: ```suggestion
### `BpkNudger`
```

The `default` bit feels redundant.

}
```

> ### Configurable `BpkConfigurableNudger`

Reviewer: ```suggestion
### `BpkConfigurableNudger`
```

'Configurable' feels redundant.

```js
import React, { Component } from 'react';
import BpkLabel from 'bpk-component-label';
import { BpkConfigurableNudger } from 'bpk-component-nudger';

class App extends Component {
  constructor() {
    super();

    this.state = {
      value: 1,
    };
  }

  handleChange = value => {
    this.setState({ value });
  };

>   compareValues = (a, b): number => {

Reviewer: ```suggestion
  compareValues = (a, b): => {
```

I _think_ we usually keep Flow types out of the example code in readmes.

Author: Sure thing, I guess this was more of a good example people could just copy and run with straight away?

Reviewer: Yeah totally! But that makes it even more of a reason to follow best practice right? If people are going to copy and use this in their own code we should make sure it's 💯

>   compareValues = (a, b): number => {

Reviewer 2: None of these use `this` so they probably shouldn't be instance methods and they definitely don't need to be bound to `this`

Author: So they shouldn't be inside the class function?

Reviewer 2: They could either be top level functions or static methods

>     const options = ['economy', 'premium', 'business', 'first'];

Reviewer: Can we abstract `options` out seeing as it's defined three times?

    const [aIndex, bIndex] = [options.indexOf(a), options.indexOf(b)];
    return aIndex - bIndex;
  };

>   incrementValue = (a): string => {

Reviewer: Can we make this a bit more explicit? We should aim to avoid variables with undescriptive names

```suggestion
  incrementValue = (currentValue): string => {
```

>     const options = ['economy', 'premium', 'business', 'first'];

Author: Could you provide some more detail about this suggestion?

>     const options = ['economy', 'premium', 'business', 'first'];

Reviewer: As in preventing accessing things outside the bound of the array, so like in this example if they didn't set a maximum but kept clicking the + button they'll eventually get `undefined` because they'll go off the end of the array. So we could put an `if` in here to stop that happening.

>     const options = ['economy', 'premium', 'business', 'first'];

Author: I don't think that scenario would ever happen as max and min are required props so it would never get over/undefined due to there always being the limit value

>     const options = ['economy', 'premium', 'business', 'first'];

Author: I am having issues seeing how a user could get the function to go out of bounds

>     const options = ['economy', 'premium', 'business', 'first'];

Reviewer: That's a good point, I didn't realise max and min were required. I'm going to defer to @k0nserv in that case

>     const options = ['economy', 'premium', 'business', 'first'];

Reviewer 2: What if there's a bug in our code 😱? It's fairly reasonable behaviour to guard it with an out of bounds check imo

    const [aIndex] = [options.indexOf(a) + 1];
    return options[aIndex];
  };

  decrementValue = (a): string => {
    const options = ['economy', 'premium', 'business', 'first'];
    const [aIndex] = [options.indexOf(a) - 1];
    return options[aIndex];
  };

  formatValue = (a): string => a.toString();

  render() {
    return (
      <div>
        <BpkLabel htmlFor="my-nudger">Number of passengers</BpkLabel>
        <BpkConfigurableNudger
          id="nudger"
          min="economy"
          max="first"
          value={this.state.value}
          onChange={this.handleChange}
          decreaseButtonLabel="Decrease"
          increaseButtonLabel="Increase"
          compareValues={this.compareValues}
          incrementValue={this.incrementValue}
          decrementValue={this.decrementValue}
          formatValue={this.formatValue}

>           inputClassName={getClassName('bpk-nudger-configurable')}

Reviewer: I'd just remove this line, `getClassName` isn't defined and we generally only document the most obvious props and not the more 'advanced' stuff like passing in class names.

| min                 | number                        | true     | -             |
| onChange            | func                          | true     | -             |
| value               | number                        | true     | -             |
| className           | string                        | false    | null          |
| buttonType          | oneOf('secondary', 'outline') | false    | secondary     |

### BpkConfigurableNudger

| Property            | PropType                      | Required | Default Value |
| ------------------- | ----------------------------- | -------- | ------------- |
| id                  | string                        | true     | -             |
| decreaseButtonLabel | string                        | true     | -             |
| increaseButtonLabel | string                        | true     | -             |
| max                 | number                        | true     | -             |
| min                 | number                        | true     | -             |
| onChange            | func                          | true     | -             |
| value               | number                        | true     | -             |

> | compareValues       | func                          | true     | -             |

Reviewer 2: We need to better document these functions. Especially `compareValues`

Author: Done

}
```

### `BpkConfigurableNudger`

```js
import React, { Component } from 'react';
import BpkLabel from 'bpk-component-label';
import { BpkConfigurableNudger } from 'bpk-component-nudger';

const options = ['economy', 'premium', 'business', 'first'];

const compareValues = (value1, value2) => {
    const [aIndex, bIndex] = [options.indexOf(value1), options.indexOf(value2)];
    return aIndex - bIndex;
  };

const incrementValue = currentValue => {
    const [aIndex] = [options.indexOf(currentValue) + 1];
    return options[aIndex];
  };

const decrementValue = currentValue => {
    const [aIndex] = [options.indexOf(currentValue) - 1];
    return options[aIndex];
  };

const formatValue = currentValue => currentValue.toString();

class App extends Component {
  constructor() {
    super();

    this.state = {
      value: 1,
    };
  }

  handleChange = value => {
    this.setState({ value });
  };

  render() {
    return (
      <div>

>         <BpkLabel htmlFor="my-nudger">Number of passengers</BpkLabel>

Reviewer: ```suggestion
        <BpkLabel htmlFor="nudger">Number of passengers</BpkLabel>
```

Wrong ID

| id                  | string                        | true     | -             |
| decreaseButtonLabel | string                        | true     | -             |
| increaseButtonLabel | string                        | true     | -             |
| max                 | number                        | true     | -             |
| min                 | number                        | true     | -             |
| onChange            | func                          | true     | -             |
| value               | number                        | true     | -             |
| compareValues       | func                          | true     | -             |
| incrementValue      | func                          | true     | -             |
| decrementValue      | func                          | true     | -             |
| formatValue         | func                          | true     | -             |
| className           | string                        | false    | null          |
| inputClassName      | string                        | false    | null          |
| buttonType          | oneOf('secondary', 'outline') | false    | secondary     |

### `compareValues`

> A comparator function similar to [Array.prototype.sort()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#Description) for comparing two values in your data set.

Reviewer 2: By convention you usually don't include the `()` when writing out methods on builtin prototypes. So `Array.prototype.sort` not `Array.prototype.sort()`

> The function is used to take either the min or max value provided and compare it to the current value to determine if the value can be incremented or decremented.

Reviewer 2: Can we mention explicitly when people should return a value less than 0, equal to zero, and greater than zero here? Saves people going to look at `Array.prototype.sort`.

### `incrementValue` & `decrementValue`

Functions that handle the incrementing or decrementing of the current selected value.

### `formatValue`

> A simple function that will allow you to set the format of the display value e.g. local dates or times.

Reviewer: These descriptions are brilliant, nice one 😄

| onChange            | func                          | true     | -             |
| value               | number                        | true     | -             |
| compareValues       | func                          | true     | -             |
| incrementValue      | func                          | true     | -             |
| decrementValue      | func                          | true     | -             |
| formatValue         | func                          | true     | -             |
| className           | string                        | false    | null          |
| inputClassName      | string                        | false    | null          |
| buttonType          | oneOf('secondary', 'outline') | false    | secondary     |

### `compareValues`

A comparator function similar to [Array.prototype.sort](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#Description) for comparing two values in your data set.

The function is used to take either the min or max value provided and compare it to the current value to determine if the value can be incremented or decremented.

> The `compareValues` requires two input values `currentValue` and `compareValue` (which will be the `min` or `max` value provided) function should return the following given each scenario.

Reviewer: The wording feels a bit clunky to me, how about:

```suggestion 
`compareValues` requires two input values: `currentValue` and `compareValue` (which will be the `min` or `max` value provided). It so return one of the following given each scenario:

* If `currentValue < compareValue`, return a negative number.
* If `currentValue == compareValue`, return zero.
* If `currentValue > compareValue`, return a positive number.
```

| decrementValue      | func                          | true     | -             |
| formatValue         | func                          | true     | -             |
| className           | string                        | false    | null          |
| inputClassName      | string                        | false    | null          |
| buttonType          | oneOf('secondary', 'outline') | false    | secondary     |

### `compareValues`

A comparator function similar to [Array.prototype.sort](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#Description) for comparing two values in your data set.

The function is used to take either the min or max value provided and compare it to the current value to determine if the value can be incremented or decremented.

`compareValues` requires two input values: `currentValue` and `compareValue` (which will be the `min` or `max` value provided). It so return one of the following given each scenario:

* If `currentValue < compareValue`, return a negative number.
* If `currentValue == compareValue`, return zero.
* If `currentValue > compareValue`, return a positive number.

> Given `currentValue` and `compareValue`:

Reviewer 2: These examples are only correct at the extremes i.e. `max` and `min`
